File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys
input = sys.stdin.readline
n, m = map(int, input().split())
graph = [[] for i in range(n)]
visited = [False] * n

for _ in range(m):
    a, b = map(int, input().rstrip().split())
    graph[a].append(b)
    graph[b].append(a)


def dfs(index, depth):
    if depth == 4:
        print(1)
        exit()
    for i in graph[index]:
        if not visited[i]:
            visited[i] = True
            dfs(i, depth + 1)
            visited[i] = False


for target in range(n):
    visited[target] = True
    dfs(target, 0)
    visited[target] = False

print(0)

File: 백준/Gold/13023. ABCDE/ABCDE.js
//문제: Doit025
//0. 문제 이해 :
//1. 시간 복잡도 확인:
//2. 적용 알고리즘 확인:
//3. 맞는 자료구조 확인:
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n");

solution(input);
function solution(input) {
    const [N, M] = input.shift().split(" ").map(Number);
    const ARR = Array.from({ length: N }, () => []);
    const VISITED = Array.from({ length: N }, () => false);
    for (let i = 0; i < M; i++) {
        const [a, b] = input[i].split(" ").map(Number);
        ARR[a].push(b);
        ARR[b].push(a);
    }
    let answer;
    const DFS = (index, depth) => {
        //DFS 재귀로 구현
        if (depth === 5) {
            answer = true;
            return;
        }
        VISITED[index] = true;
        ARR[index].forEach((e) => {
            if (!VISITED[e]) {
                DFS(e, depth + 1);
            }
        });
        VISITED[index] = false;
    };
    for (let i = 0; i < N; i++) {
        DFS(i, 1);
        if (answer) break;
    }
    if (answer) console.log(1);
    else console.log(0);
}
